### PR TITLE
will persist snippet data across the login session

### PR DESCRIPTION
**Pull request by Niharika0104**
Fixes - #399

### CODE REVIEW

File: src/hooks/use-global-store.ts [hunk 1]
```diff
@@ -1,13 +1,22 @@
+import { AnyCircuitElement } from "circuit-json"
 import { create } from "zustand"
 import { persist } from "zustand/middleware"
 
+interface snippetInfo {
+  circuitJson?: AnyCircuitElement[] | null
+  code: string
+  snippetType: "board" | "package" | "model" | "footprint" | string | undefined
```
Comment (imrishabh18) [inline src/hooks/use-global-store.ts:8]: The snippet type shouldn't be `undefined` or `string`, when we are using the const names. It defeats the point of it

File: src/hooks/use-run-tsx/index.tsx [hunk 2]
```diff
@@ -26,7 +26,7 @@ export const useRunTsx = ({
 }: {
   code?: string
   userImports?: Record<string, object>
-  type?: "board" | "footprint" | "package" | "model"
+  type?: "board" | "footprint" | "package" | "model" | string
```
Comment (imrishabh18) [inline src/hooks/use-run-tsx/index.tsx:29]: Here as well, we shouldn't be using `string`

**Review by imrishabh18 — CHANGES_REQUESTED** [review]

**Comment by Niharika0104** [comment]
@imrishabh18 Okay sure.So we should split tests across multiple files ?

**Comment by Niharika0104** [comment]
@imrishabh18 i'm unable to test these changes locally coz locally we are using fakeApi and not the redirecting the user to the github oauth page
image
the user will click on login to save button and when in prod env it will redirect user to the github oauth page and from there to the authorize page and if the authorize page sees that guest snippet variable exists in the local storage it will redirected to the editor page and I will extract the previous edited info from guest snippet and set it . If that's good then i will start writing a test for this

**Comment by imrishabh18** [comment]
@Niharika0104 Yeah, correct. I think we can merge this without test. What do you think @seveibar?

On the UI, you have converted the `Not logged in, can't save` tag beside the save button to `Login to save` as a button. From the user point of view, that button appears to be a tag, which is pointing me to Login on the header and not click on the tag itself.

**Comment by Niharika0104** [comment]
@imrishabh18 yeah when we click on login to save it will redirect to GitHub oauth page and on entering their credentials it will redirect them to editor page and the code editor and manual edits  json file will be populated with the info we saved in guest snippet variable present in the local storage
Hope this makes sense!!

File: src/components/CodeAndPreview.tsx [hunk 3]
```diff
@@ -50,14 +52,19 @@ export function CodeAndPreview({ snippet }: Props) {
...
-    snippet?.snippet_type ?? (templateFromUrl.type as any)
+  const [snippetType, setSnippetType] = useState<
+    "board" | "package" | "model" | "footprint"
+  >(snippet?.snippet_type ?? (templateFromUrl.type as any))
 
   useEffect(() => {
     if (snippet?.code) {
       setCode(snippet.code)
       setLastRunCode(snippet.code)
+    } else if (!isLoggedIn && guestSnippet) {
```
Comment (seveibar) [inline src/components/CodeAndPreview.tsx:63]: i think this breaks the loading of templates?

**Comment by seveibar** [comment]
@Niharika0104 I think this should be still testable in playwright if you create a fake-api login route, the url is something like:

- `internal/oauth/github/authorize?next=${nextUrl}/authorize`

Overall i like the implementation, except i think most users will click "login" in the header rather than the badge.

It's a bit of work but if you're down i would recommend pausing this PR, then implementing a PR that just allows logging in inside playwright tests (issue: https://github.com/tscircuit/snippets/issues/409)

Overall it's a good idea to store the guest session inside the global store. But I'm cautious about the loading process- we might want to have a specialized hook or something that runs the appropriate state machine because the `<CodeAndPreview />` file scares me. But I'm not quite sure how to do this refactor yet-

**Comment by seveibar** [comment]
@Niharika0104 i was looking at the logic in a bit more detail, it seems like the playwright test should be able to login due to this login logic:

image

**Comment by Niharika0104** [comment]
Yeah the tests are fine now but when we click on login /sign up on the header it takes us to the home page(/) but when we click on the login to save button it will take us to GitHub oauth and then to the editor page.
I tried testing it in the vercel preview site but GitHub oauth is not working there @seveibar

**Comment by seveibar** [comment]
Yea i cant enable auth in preview

We should use the same code/logic for both logins, there's a sign in hook that should handle it iirc

**Comment by Niharika0104** [comment]
Yeah Im aware of that hook let me try again 
Thanks @seveibar
